src/acceleration_structure.h

```diff
@@ -0,0 +1,251 @@
+  BLAS* GetUsedBLASPtr() const { return used_blas_ptr_; }
+
+  void SetTransform(const std::vector<float>& transform) {
+    transform_ = transform;
+  }
+  const float* GetTransform() const { return transform_.data(); }
+
+  void SetInstanceIndex(uint32_t instanceCustomIndex) {
+    instanceCustomIndex_ = instanceCustomIndex;
+    assert(instanceCustomIndex_ == instanceCustomIndex);
+  }
+  uint32_t GetInstanceIndex() const { return instanceCustomIndex_; }
+
+  void SetMask(uint32_t mask) {
+    mask_ = mask;
+    assert(mask_ == mask);
+  }
+  uint32_t GetMask() const { return mask_; }
+
+  void SetOffset(uint32_t offset) {
+    instanceShaderBindingTableRecordOffset_ = offset;
+    assert(instanceShaderBindingTableRecordOffset_ == offset);
+  }
+  uint32_t GetOffset() const { return instanceShaderBindingTableRecordOffset_; }
+
+  void SetFlags(uint32_t flags) {
+    flags_ = flags;
+    assert(flags_ == flags);
+  }
+  uint32_t GetFlags() const { return flags_; }
+
+ private:
+  std::string used_blas_name_;
+  BLAS* used_blas_ptr_;
+  std::vector<float> transform_;
+  uint32_t instanceCustomIndex_ : 24;
+  uint32_t mask_ : 8;
+  uint32_t instanceShaderBindingTableRecordOffset_ : 24;
+  uint32_t flags_ : 8;
+};
+
+class TLAS {
+ public:
+  TLAS();
+  ~TLAS();
+
+  void SetName(const std::string& name) { name_ = name; }
+  std::string GetName() const { return name_; }
+
+  void AddInstance(std::unique_ptr<BLASInstance> instance) {
+    blas_instances_.push_back(std::shared_ptr<BLASInstance>(instance.release()));
+  }
+  size_t GetInstanceSize() { return blas_instances_.size(); }
+  std::shared_ptr<BLASInstance>* GetInstanceData() {
+    return blas_instances_.data();
+  }
+  std::vector<std::shared_ptr<BLASInstance>>& GetInstances() {
+    return blas_instances_;
+  }
+
+ private:
+  std::string name_;
+  std::vector<std::shared_ptr<BLASInstance>> blas_instances_;
+};
+
+class ShaderGroup {
+ public:
+  ShaderGroup();
+  ~ShaderGroup();
+
+  void SetName(const std::string& name) { name_ = name; }
+  std::string GetName() const { return name_; }
+
+  void SetGeneralShader(Shader* shader) { generalShader_ = shader; }
+  Shader* GetGeneralShader() const { return generalShader_; }
+
+  void SetClosestHitShader(Shader* shader) {
+    closestHitShader_ = shader;
+  }
+  Shader* GetClosestHitShader() const { return closestHitShader_; }
+
+  void SetAnyHitShader(Shader* shader) { anyHitShader_ = shader; }
+  Shader* GetAnyHitShader() const { return anyHitShader_; }
+
+  void SetIntersectionShader(Shader* shader) {
+    intersectionShader_ = shader;
+  }
+  Shader* GetIntersectionShader() const { return intersectionShader_; }
+
+  bool IsGeneralGroup() const { return generalShader_ != nullptr; }
+  bool IsHitGroup() const {
+    return closestHitShader_ != nullptr || anyHitShader_ != nullptr ||
+           intersectionShader_ != nullptr;
+  }
+
+  std::string name_;
+  Shader* generalShader_;
```

Comment:
Vulkan spec on `VkRayTracingShaderGroupCreateInfoKHR` :

> generalShader is the index of the ray generation, miss, or callable shader from VkRayTracingPipelineCreateInfoKHR::pStages in the group if the shader group has type of VK_RAY_TRACING_SHADER_GROUP_TYPE_GENERAL_KHR, and VK_SHADER_UNUSED_KHR otherwise.

Thus no rename.